hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -201,15 +208,51 @@ private void handlePlatformTransaction(
             recordBuilder.status(ResponseCodeEnum.UNKNOWN);
         }
 
-        // TODO: handle long scheduled transactions
+        // store all records at once
+        final var recordListResult = recordListBuilder.build();
+        recordManager.endUserTransaction(recordListResult.recordStream());
+
+        // Verify if the transaction is a duplicate and add it to the cache
+        if (preHandleResult != null) { // this should always be true, checking just in case
+            try {
+                checkDuplicatesAndIncludeInCache(
+                        preHandleResult,
+                        recordListResult.mainRecord().recordStreamItem().record(),
+                        consensusNow);
+            } catch (final PreCheckException e) {
+                recordFailedTransaction(e.responseCode(), recordBuilder, recordListBuilder);
+            }
+        }
 
         // TODO: handle system tasks
+    }
 
-        // store all records at once
-        recordManager.endUserTransaction(recordListBuilder.build());
+    private void checkDuplicatesAndIncludeInCache(
+            @NonNull final PreHandleResult preHandleResult,
+            @NonNull final TransactionRecord transactionRecord,
+            @NonNull final Instant consensusNow)
+            throws PreCheckException {
+        final var transactionID = transactionRecord.transactionIDOrThrow();
+        final var foundTransactionRecord = hederaRecordCache.getRecord(transactionID);
+
+        addTransactionToCache(preHandleResult, transactionRecord, consensusNow);
+
+        if (foundTransactionRecord != null) {
+            throw new PreCheckException(DUPLICATE_TRANSACTION);
+        }
+    }
+
+    private void addTransactionToCache(
+            @NonNull final PreHandleResult preHandleResult,
+            @NonNull final TransactionRecord transactionRecord,
+            @NonNull final Instant consensusNow) {
+        final var txBody = preHandleResult.txInfo().txBody();
+        final var nodeId = txBody.nodeAccountID().accountNum();
```

Comment:
The nodeId and the node's account ID are two different things. The node ID uniquely identifies the node over the lifetime of the node, whereas (theoretically) the operator account tied to the node can change over time.

Reply:
I think we need to add the nodeId to the PreHandleResult. The PreHandleWorkflow knows what the node ID is, so it just needs to stash it away so you can pick it up in the handle workflow.

Reply:
Sounds good, but I'm not sure how to get the nodeID from the PreHandleWorkflow.
We were looking it with Miro and can we use the networkInfo class to get the nodeID? Something like:
`networkInfo.selfNodeInfo().accountId().accountNum()`

Reply:
@rbair23 As we discussed in the meeting yesterday, I've tried passing the nodeId to the preHandle but the underlying problem is `HederaWorkflow.getCurrentPreHandleResult()`. Since we don't know if a pre-handle is called before the handle there is no way for me to access the nodeId passed on pre-handle in the case where the pre-handle was not run yet. This is the case after this if statement(`if (preHandleStillValid(configuration, metadata))`) in the getCurrentPreHandleResult method.

I've noticed that we are building the **creator** in `getCurrentPreHandleResult` the same way we build it in `Hedera` when we pass it - `final var creator = networkInfo.nodeInfo(platformEvent.getCreatorId().id());`
If that's the case the `platformEvent.getCreatorId().id()` should be the **nodeId** and I can use that directly in `HandleWorkflow` without the need to pass it from `Hedera`. What do you think?